### PR TITLE
[KlaudSol-CMS-#137] -Add pagination and page limit to tables

### PR DESCRIPTION
**Pull request by Norlant1**
1. created a customized and reusable Pagination component `<AppContentPagination/>`. The component receives 8 required and 1 optional properties.

1 pagination
1-1 pagination props

note: default number of entries is temporarily set to 1 in the video in order to show case `next page`, `previous page`, `min page`, and `max page` functionality. The default number of entries is set to 10 in PR repository.

https://user-images.githubusercontent.com/109654448/215252353-856a9fa7-c6fc-420a-b356-1d6baa7fa693.mp4

2. refactored backend queries and added limit and offset as the function's parameters.

2 refactored backend

3. only show the skeleton table in the first load.

3 skeleton

4. abort all previous api request when the user moves to another page or different slug to avoid flickering of all requested data as shown in the video below.

WITHOUT ABORT CONTROLLER

https://user-images.githubusercontent.com/109654448/215252420-b32f613e-efef-4fdf-888a-2488da118239.mp4

WITH ABORT CONTROLLER

https://user-images.githubusercontent.com/109654448/215252437-4c959405-e5fc-4668-989c-3226c1a34ee4.mp4

application of code (abort)

4 abort

4-2 abort

5. created a separate file for const variables and reducer.

### CODE REVIEW

**Comment by Norlant1** [comment]
updated.

1. removed buttons' shadows

File: backend/models/core/Entity.js [hunk 1]
```diff
@@ -169,6 +169,101 @@ class Entity {
         })); 
   }
 
+  static async findByPageAndEntry({entity_type_slug , entry, page }) {
+    const db = new DB();
+    
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:175]: Kindly remove excess lines of code (Line 175 - 176)

File: backend/models/core/Entity.js [hunk 2]
```diff
@@ -169,6 +169,101 @@ class Entity {
...
 
+  static async findByPageAndEntry({entity_type_slug , entry, page }) {
+    const db = new DB();
+    
+   
+    
+    let totalRows;
+    let totalOrders;
+    
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:180]: Remove excess line (LIne 180)

File: backend/models/core/Entity.js [hunk 3]
```diff
@@ -169,6 +169,101 @@ class Entity {
...
+                          WHERE entity_types.slug = :entity_type_slug;
+                          `;
+
+    const totalRowsData = await db.executeStatement(totalRowsSQL,[
+      {name: 'entity_type_slug', value:{stringValue: entity_type_slug}},
+    ]);
+
+    [{longValue:totalRows}] = totalRowsData.records[0]
+
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:192]: Remove excess line (Line 192)

File: backend/models/core/Entity.js [hunk 4]
```diff
@@ -169,6 +169,101 @@ class Entity {
...
+    LEFT JOIN attributes ON attributes.entity_type_id = entity_types.id
+    WHERE entity_types.slug = :entity_type_slug`;
+
+    const totalOrdersData = await db.executeStatement(totalOrdersSQL,[
+      {name: 'entity_type_slug', value:{stringValue: entity_type_slug}},
+    ]);
+    [{longValue:totalOrders}] = totalOrdersData.records[0]
+    
+    
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:204]: Remove excess line. (Lines 203 - 204)

File: backend/models/core/Entity.js [hunk 5]
```diff
@@ -169,6 +169,101 @@ class Entity {
...
+    let limit = entry ? (( totalOrders / totalRows ) * entry ) : 10;
+    let offset = page ?  limit * page : 0;
+
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:208]: Remove excess line (Line 208)

File: backend/models/core/Entity.js [hunk 6]
```diff
@@ -169,6 +169,101 @@ class Entity {
...
+                    entity_types.slug = :entity_type_slug  
+                ORDER BY entities.id, attributes.\`order\` ASC
+                LIMIT ${limit} OFFSET ${offset}
+                `;
+              
+    const dataRaw = await db.executeStatement(sql, [
+        {name: 'entity_type_slug', value:{stringValue: entity_type_slug}},
+    ]);
+    
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:230]: Remove excess line (Line 230)

File: backend/models/core/Entity.js [hunk 7]
```diff
@@ -169,6 +169,101 @@ class Entity {
...
+        value_string,
+        value_long_string,
+        value_integer,
+        value_datetime,
+        value_double
+      })); 
+
+      return {total_rows:totalRows, data}
+}
+
```
Comment (angeli-mercedes) [inline backend/models/core/Entity.js:263]: Remove excess lines (264 - 266)

File: components/elements/inner/ContentBuilderSubMenu.js [hunk 8]
```diff
@@ -1,154 +1,142 @@
...
+      { title: "Test", id: 4 },
+      { title: "Test", id: 5 },
+    ],
+    selectedType: 1,
+    isLoading: false,
+  };
+
+  const SET_SELECTED_TYPE = "SET_SELECTED_TYPE";
+  const LOADING = "LOADING";
+  const reducer = (state, action) => {
```
Comment (angeli-mercedes) [inline components/elements/inner/ContentBuilderSubMenu.js:36]: Kindly refactor this to a separate file.

File: components/elements/inner/ContentManagerSubMenu.js [hunk 9]
```diff
@@ -1,90 +1,103 @@
...
-    
-    const [state, dispatch] = useReducer(reducer, initialState);
-    
-    const onStartLoad = () => {
-      //only display on first load
-      if(!rootState.entityTypesHash) {
-        dispatch({type: LOADING, payload: true});
-      }
-    };
+  const SET_SELECTED_TYPE = "SET_SELECTED_TYPE";
```
Comment (angeli-mercedes) [inline components/elements/inner/ContentManagerSubMenu.js:34]: Kindly refactor this to a separate file.

File: components/klaudsolcms/pagination/contentPaginationVariables.js [hunk 10]
```diff
@@ -0,0 +1,4 @@
+export const maximumNumberOfPage = 10;
```
Comment (angeli-mercedes) [inline components/klaudsolcms/pagination/contentPaginationVariables.js:1]: Kindly move this file to constants folder and export it to constants folder > index.js

File: components/reducers/entitiesReducer.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  setsRenderer:0,
+  firstFetch:true,
+};
+
+
+export const LOADING = "LOADING";
+export const REFRESH = "REFRESH";
+export const CLEANUP = "CLEANUP";
+
+export const SET_VALUES = "SET_VALUES";
```
Comment (angeli-mercedes) [inline components/reducers/entitiesReducer.js:20]: You can refactor this to a separate file. I suggest you add this to reducers > actions.js like this

Screenshot 2023-02-06 at 1 12 07 PM

File: components/reducers/entitiesReducer.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+export const SET_COLUMNS = "SET_COLUMNS";
+export const SET_ENTITY_TYPE_NAME = "SET_ENTITY_TYPE_NAME";
+export const SET_ROWS = "SET_ROWS";
+export const SET_ENTRIES = "SET_ENTRIES";
+export const SET_PAGE = "SET_PAGE";
+export const ROWS_SET = "ROWS_SET";
+export const PAGE_SETS_RENDERER = "PAGE_SETS_RENDERER";
+export const SET_FIRST_FETCH = "SET_FIRST_FETCH";
+
+export const reducer = (state, action) => {
```
Comment (angeli-mercedes) [inline components/reducers/entitiesReducer.js:30]: Kindly refactor this to a separate reducers file.

**Review by angeli-mercedes — COMMENTED** [review]
Kindly see the comments, fix the merge conflicts and assign this PR to yourself. Thanks!

**Review by angeli-mercedes — CHANGES_REQUESTED** [review]
Kindly address the comments.

**Pull request closed without merging**